workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System;
+using System.Collections.Generic;
+using Improbable.Gdk.Core.GameObjectRepresentation;
+using Improbable.Worker;
+using Unity.Entities;
+using UnityEngine;
+using Entity = Unity.Entities.Entity;
+
+namespace Improbable.Gdk.Core.MonoBehaviours
+{
+    public abstract class ReaderBase<TSpatialComponentData, TComponentUpdate>
+        : IReader<TSpatialComponentData, TComponentUpdate>,
+            IReaderInternal
+        where TSpatialComponentData : ISpatialComponentData
+        where TComponentUpdate : ISpatialComponentUpdate
+    {
+        protected readonly Entity Entity;
+        protected readonly EntityManager EntityManager;
+
+        private readonly ILogDispatcher logDispatcher;
+
+        protected ReaderBase(Entity entity, EntityManager entityManager, ILogDispatcher logDispatcher)
+        {
+            Entity = entity;
+            EntityManager = entityManager;
+            this.logDispatcher = logDispatcher;
+        }
+
+        public Authority Authority
+        {
+            get
+            {
+                if (EntityManager.HasComponent<AuthorityLossImminent<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.AuthorityLossImminent;
+                }
+
+                if (EntityManager.HasComponent<Authoritative<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.Authoritative;
+                }
+
+                if (EntityManager.HasComponent<NotAuthoritative<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.NotAuthoritative;
+                }
+
+                throw new ReaderAuthorityGetFailedException("No authority component found for this entity.");
```

Comment:
Though not particularly useful, how about adding the entityindex of the entity in the error message?